publish-build.sh

```diff
@@ -17,51 +17,68 @@ set -x # print all commands
 shopt -s nullglob globstar
 # end: shellharden
 
+function printerr () {
+    >&2 echo -e "\e[91m${1}\e[0m"
+}
+
 if [[ ! ${TRAVIS_TAG:-} ]]; then
-    echo "Not a built for a Git tag, do not publish!"
+    printerr "Not built from a Git tag, do not publish!"
     exit 0
 fi
 
+DRYRUN=0
+if [ $# -gt 0 ] && [ "$1" == "--dry-run" ]; then
+    DRYRUN=1
+    printerr "PERFORMING DRY RUN"
+fi
+
 # this part needs to be expanded for other dist-tags
 # based on api levels
 
-function publishPackage () {
-    local COMPONENT=$1
-    local REPO_DIR=$2
-
-    COMPONENT=${COMPONENT//_/-}
+function exec () {
+    local CMD=$1
 
-    BUILD_REPO_NAME="${COMPONENT}"
-    BUILD_REPO_DIR="tmp/${BUILD_REPO_NAME}"
-
-    version=$(node -pe "require('./package.json').version")
-    echo "Publishing version: ${version}"
+    if [ $DRYRUN -eq 1 ]; then
+        echo $CMD
+    else
+        `${CMD}`
+    fi
+}
 
+function publishPackage () {
+    local PACKAGE_DIR=$1
+    local PACKAGE_JSON="${PACKAGE_DIR}package.json"
 
-    if [[ -d "$BUILD_REPO_DIR" ]]; then
-        npm publish "$BUILD_REPO_DIR" --tag "$DIST_TAG" --access public
+    if [[ ! -e ${PACKAGE_JSON} ]]; then
+        printerr "Package.json file '${PACKAGE_JSON}' does not exist, skipping publish."
     else
-        npm publish --tag "$DIST_TAG" --access public
+        name=$(node -pe "require('${PACKAGE_DIR}package.json').name")
+        version=$(node -pe "require('${PACKAGE_DIR}package.json').version")
+        echo "Publishing package: ${name} @ ${version}"
+
+        exec "npm publish \"$PACKAGE_DIR\" --tag \"$DIST_TAG\" --access public"
     fi
 }
 
 DIST_TAG=latest
-
-ROOT=$(basename $(pwd))
+BUILDS_DIR="./tmp/builds"
 
 echo "//registry.npmjs.org/:_authToken=${NPM_TOKEN}" > ~/.npmrc
 echo "//registry.npmjs.org/:username=travis4dhis2" >> ~/.npmrc
 echo "//registry.npmjs.org/:email=deployment@dhis2.org" >> ~/.npmrc
 
-if [[ ! -d "packages" ]]; then
+if [ ! -d "./packages" ] && [ ! -d "${BUILDS_DIR}" ]; then
```

Comment:
Using the single-brackets the way you have done is correct, but what's the reason you went for single brackets here (`sh` compliant conditional) when the script itself assumes `bash`?

Using the `bash`-style double-brackets is a bit safer for developers who aren't familiar with the differences between the two.

Reply:
Habit.  Changed to use double brackets.